pandas/tests/groupby/test_groupby.py

```diff
@@ -2012,3 +2012,48 @@ def test_groups_repr_truncates(max_seq_items, expected):
 
         result = df.groupby(np.array(df.a)).groups.__repr__()
         assert result == expected
+
+
+def test_sorted_missing_category_values():
+    # GH 28597
+    df = pd.DataFrame(
+        {
+            "foo": [
+                "small",
+                "large",
+                "large",
+                "large",
+                "medium",
+                "large",
+                "large",
+                "medium",
+            ],
+            "bar": ["C", "A", "A", "C", "A", "C", "A", "C"],
+        }
+    )
+    df["foo"] = (
+        df["foo"]
+        .astype("category")
+        .cat.set_categories(["tiny", "small", "medium", "large"], ordered=True)
+    )
+
+    expected = pd.DataFrame(
+        {
+            "tiny": {"A": 0, "C": 0},
+            "small": {"A": 0, "C": 1},
+            "medium": {"A": 1, "C": 1},
+            "large": {"A": 3, "C": 2},
+        }
+    )
+    expected = expected.rename_axis("bar", axis="index")
+    expected.columns = pd.CategoricalIndex(
+        ["tiny", "small", "medium", "large"],
+        categories=["tiny", "small", "medium", "large"],
+        ordered=True,
+        name="foo",
+        dtype="category",
+    )
+
+    result = df.groupby(["bar", "foo"]).size().unstack()
```

Comment:
I _think_ the change in behaviour that needs to be tested is the size method and the unstack call is not relevant in the issue OP.

Reply:
There are two separate issues that have been fixed, and I think that this tests them both. One is that missing categorical values in `groupby` would cause the categories to become unsorted https://github.com/pandas-dev/pandas/issues/28597 . That is the one that I have mentioned in the commit message. 

The other is that previously, empty groups would be missing from the output even when grouping with categories https://github.com/pandas-dev/pandas/issues/23865, which I think this test also catches as a side effect.

Reply:
IIUC it's the size call that is the root cause of the issues and this issue was resolved in #29690

> which I think this test also catches as a side effect

maybe one of the issues should be closed as a duplicate?